### PR TITLE
NO-JIRA:Use expressive Gomega matchers with Eventually

### PR DESCRIPTION
**Pull request by nrb**
The matchers used with `Eventually` checks boiled down to a boolean
comparison, which is often unhelpful when debugging. This change uses
more detailed Gomega matchers in order to aid developers making changes.

This explicitly does _not_ try to consolidated the inconsistent import usages of Gomega within this repository.

### CODE REVIEW

File: pkg/controllers/cloud_config_sync_controller_test.go [hunk 1]
```diff
@@ -224,10 +224,12 @@ var _ = Describe("Cloud config sync controller", func() {
...
 		if err == nil || !apierrors.IsNotFound(err) {
-			Eventually(func() bool {
-				err := cl.Delete(context.Background(), co)
-				return err == nil || apierrors.IsNotFound(err)
-			}).Should(BeTrue())
+			Eventually(func() error {
+				return cl.Delete(context.Background(), co)
+			}).Should(SatisfyAny(
+				Not(HaveOccurred()),
+				MatchError(apierrors.IsNotFound, "IsNotFound"),
```
Comment (JoelSpeed) [inline pkg/controllers/cloud_config_sync_controller_test.go:231]: Not seen this used like this before, but cool, checked the docs and it works, nice find

File: pkg/controllers/cloud_config_sync_controller_test.go [hunk 2]
```diff
@@ -241,8 +243,8 @@ var _ = Describe("Cloud config sync controller", func() {
 		for _, cm := range allCMs.Items {
 			Expect(cl.Delete(ctx, cm.DeepCopy(), deleteOptions)).To(Succeed())
 			Eventually(
-				apierrors.IsNotFound(cl.Get(ctx, client.ObjectKeyFromObject(cm.DeepCopy()), &corev1.ConfigMap{})),
-			).Should(BeTrue())
+				cl.Get(ctx, client.ObjectKeyFromObject(cm.DeepCopy()), &corev1.ConfigMap{}),
```
Comment (JoelSpeed) [inline pkg/controllers/cloud_config_sync_controller_test.go:246]: This is a gotcha of using eventually statements, what you're doing here is calling `cl.Get` while constructing the `Eventually`, which means `Eventually` gets a single result, which never changes.

The original is also broken, but you have two choices here to fix it:
```suggestion
				func() error { return cl.Get(ctx, client.ObjectKeyFromObject(cm.DeepCopy()), &corev1.ConfigMap{})},
```
Or use [komega](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/envtest/komega) as we do in CPMS
```suggestion
				komega.Get(ctx, client.ObjectKeyFromObject(cm.DeepCopy()), &corev1.ConfigMap{}),
```
You'll need to either create a new instance of komega or set the global komega client before you can do this.

File: pkg/controllers/cloud_config_sync_controller_test.go [hunk 3]
```diff
@@ -251,16 +253,15 @@ var _ = Describe("Cloud config sync controller", func() {
 		infra := makeInfrastructureResource(configv1.AzurePlatformType)
 		Expect(cl.Delete(ctx, infra)).To(Succeed())
 		Eventually(
-			apierrors.IsNotFound(cl.Get(ctx, client.ObjectKeyFromObject(infra), infra)),
-		).Should(BeTrue())
+			cl.Get(ctx, client.ObjectKeyFromObject(infra), infra),
```
Comment (JoelSpeed) [inline pkg/controllers/cloud_config_sync_controller_test.go:256]: Same gotcha as above

File: pkg/controllers/cloud_config_sync_controller_test.go [hunk 4]
```diff
@@ -251,16 +253,15 @@ var _ = Describe("Cloud config sync controller", func() {
...
 
 		networkResource := makeNetworkResource()
 		Expect(cl.Delete(ctx, networkResource)).To(Succeed())
 
-		Eventually(func() bool {
-			err := cl.Get(ctx, client.ObjectKeyFromObject(networkResource), networkResource)
-			return apierrors.IsNotFound(err)
-		}).Should(BeTrue(), "Expected not found error")
+		Eventually(
+			cl.Get(ctx, client.ObjectKeyFromObject(networkResource), networkResource),
```
Comment (JoelSpeed) [inline pkg/controllers/cloud_config_sync_controller_test.go:263]: Same gotcha as above

File: pkg/controllers/cloud_config_sync_controller_test.go [hunk 5]
```diff
@@ -463,10 +462,12 @@ var _ = Describe("Cloud config sync reconciler", func() {
 		co := &configv1.ClusterOperator{}
 		err := cl.Get(context.Background(), client.ObjectKey{Name: clusterOperatorName}, co)
 		if err == nil || !apierrors.IsNotFound(err) {
-			Eventually(func() bool {
-				err := cl.Delete(context.Background(), co)
-				return err == nil || apierrors.IsNotFound(err)
-			}).Should(BeTrue())
+			Eventually(
+				cl.Delete(context.Background(), co),
```
Comment (JoelSpeed) [inline pkg/controllers/cloud_config_sync_controller_test.go:466]: Wrapping gotcha as above

File: pkg/controllers/cloud_config_sync_controller_test.go [hunk 6]
```diff
@@ -478,24 +479,23 @@ var _ = Describe("Cloud config sync reconciler", func() {
 		// omitted error intentionally, 404 might be there for some cases
 		cl.Delete(ctx, infra) //nolint:errcheck
 		Eventually(
-			apierrors.IsNotFound(cl.Get(ctx, client.ObjectKeyFromObject(infra), infra)),
-		).Should(BeTrue())
+			cl.Get(ctx, client.ObjectKeyFromObject(infra), infra),
```
Comment (JoelSpeed) [inline pkg/controllers/cloud_config_sync_controller_test.go:482]: Wrapping gotcha

File: pkg/controllers/cloud_config_sync_controller_test.go [hunk 7]
```diff
@@ -478,24 +479,23 @@ var _ = Describe("Cloud config sync reconciler", func() {
...
 
 		networkResource := makeNetworkResource()
 		Expect(cl.Delete(ctx, networkResource)).To(Succeed())
 
-		Eventually(func() bool {
-			err := cl.Get(ctx, client.ObjectKeyFromObject(networkResource), networkResource)
-			return apierrors.IsNotFound(err)
-		}).Should(BeTrue(), "Expected not found error")
+		Eventually(
+			cl.Get(ctx, client.ObjectKeyFromObject(networkResource), networkResource),
```
Comment (JoelSpeed) [inline pkg/controllers/cloud_config_sync_controller_test.go:489]: Wrapping gotcha

File: pkg/controllers/cloud_config_sync_controller_test.go [hunk 8]
```diff
@@ -478,24 +479,23 @@ var _ = Describe("Cloud config sync reconciler", func() {
...
+		).Should(MatchError(apierrors.IsNotFound, "IsNotFound"))
 
 		allCMs := &corev1.ConfigMapList{}
 		Expect(cl.List(ctx, allCMs)).To(Succeed())
 		for _, cm := range allCMs.Items {
 			Expect(cl.Delete(ctx, cm.DeepCopy(), deleteOptions)).To(Succeed())
 			Eventually(
-				apierrors.IsNotFound(cl.Get(ctx, client.ObjectKeyFromObject(cm.DeepCopy()), &corev1.ConfigMap{})),
-			).Should(BeTrue())
+				cl.Get(ctx, client.ObjectKeyFromObject(cm.DeepCopy()), &corev1.ConfigMap{}),
```
Comment (JoelSpeed) [inline pkg/controllers/cloud_config_sync_controller_test.go:497]: Wrapping gotcha

File: pkg/controllers/cloud_config_sync_controller_test.go (hunk 1 above)
Comment (nrb) [reply pkg/controllers/cloud_config_sync_controller_test.go:231]: Might be new. I'm a little surprised there's not a matcher already defined that does this same thing, but I wasn't able to find one.

**Pull request merged**